crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -479,962 +673,117 @@ impl CheckpointExecutor {
             }
         }
 
-        if let Some(checkpoint_data) = checkpoint_data {
-            self.commit_index_updates_and_enqueue_to_subscription_service(checkpoint_data)
-                .await;
+        if let Some(checkpoint_data) = ckpt_state.full_data.take() {
+            self.commit_index_updates_and_enqueue_to_subscription_service(checkpoint_data);
         }
 
-        if !checkpoint.is_last_checkpoint_of_epoch() {
+        self.accumulator
+            .accumulate_running_root(&self.epoch_store, seq, ckpt_state.accumulator)
+            .await
+            .expect("Failed to accumulate running root");
+
+        if ckpt_state.data.checkpoint.is_last_checkpoint_of_epoch() {
+            self.checkpoint_store
+                .insert_epoch_last_checkpoint(self.epoch_store.epoch(), &ckpt_state.data.checkpoint)
+                .expect("Failed to insert epoch last checkpoint");
+
             self.accumulator
-                .accumulate_running_root(epoch_store, checkpoint.sequence_number, checkpoint_acc)
-                .await
-                .expect("Failed to accumulate running root");
-            self.bump_highest_executed_checkpoint(checkpoint);
+                .accumulate_epoch(self.epoch_store.clone(), seq)
+                .expect("Accumulating epoch cannot fail");
+
+            self.checkpoint_store
+                .prune_local_summaries()
+                .tap_err(|e| debug_fatal!("Failed to prune local summaries: {}", e))
+                .ok();
         }
+
+        fail_point!("crash");
+
+        self.bump_highest_executed_checkpoint(&ckpt_state.data.checkpoint);
     }
 
     /// If configured, commit the pending index updates for the provided checkpoint as well as
     /// enqueuing the checkpoint to the subscription service
-    async fn commit_index_updates_and_enqueue_to_subscription_service(
-        &self,
-        checkpoint: CheckpointData,
-    ) {
+    fn commit_index_updates_and_enqueue_to_subscription_service(&self, checkpoint: CheckpointData) {
         if let Some(rpc_index) = &self.state.rpc_index {
             rpc_index
                 .commit_update_for_checkpoint(checkpoint.checkpoint_summary.sequence_number)
                 .expect("failed to update rpc_indexes");
         }
 
         if let Some(sender) = &self.subscription_service_checkpoint_sender {
-            if let Err(e) = sender.send(checkpoint).await {
-                tracing::warn!("unable to send checkpoint to subscription service: {e}");
+            if let Err(e) = sender.blocking_send(checkpoint) {
+                warn!("unable to send checkpoint to subscription service: {e}");
```

Comment:
Why this change? given `commit_index_updates_and_enqueue_to_subscription_service` is still called from an async function this will panic since its called from an async context: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Sender.html#method.blocking_send

Reply:
If instead you want to wrap the call to `commit_update_for_checkpoint` in a spawn_blocking that may have been what you were trying to accomplish? (or alternatively calling this function from a spawn_blocking would also work)

Reply:
Good catch! I guess there aren't tests for this after all? In any case, yes, I did mean to put the calling code inside a spawn_blocking

Reply:
there should be a test for this, i did write one! I wonder if it didn't run or something else...

Reply:
Still need to move this into a spawn_blocking call?

Reply:
so there is a test: 

```
PASS [  25.125s] sui-e2e-tests::rpc checkpoints::subscribe_checkpoint
```

but i'm not sure why its passing when i think this should panic? this passing test is from the simtest run so i wonder if something is different in that environment?

Reply:
🤦🏼‍♂️ ok i see that is was actually properly wrapped in a spawn_blocking call...I completely missed this sorry!

Reply:
I had not wrapped it in spawn_blocking at the time of your original comment. you're not crazy!